rust/src/tree.rs

```diff
@@ -0,0 +1,214 @@
+                // TODO: moving below `let child =` does not work. Why?
+                if prefix_len == 0 {
+                    continue;
+                }
+
+                let child = &mut curr.children[i];
+                
+                if prefix_len == child.ids.len() {
+
+                    // move to child
+                    curr = child;
+                    curr.count += 1;
+                    curr_idx += prefix_len;
+
+                    has_match = true;
+
+                    break;
+                } else {
+
+                    // split child
+                    let new_child = Node {
+                        children: child.children.clone(), // move the owndership of child.children
+                        ids: child.ids[prefix_len..].to_vec(),
+                        count: child.count,
+                    };
+
+                    child.ids = child.ids[..prefix_len].to_vec();
+                    child.children = vec![new_child];
+
+                    curr = child;
+                    curr.count += 1;
+                    curr_idx += prefix_len;
+
+                    has_match = true;
+
+                    break;
+                }
+
+            }
+
+            if !has_match {
+                // create new child
+                let new_child = Node {
+                    children: Vec::new(),
+                    ids: input_ids[curr_idx..].to_vec(),
+                    count: 0,
+                };
+
+                curr.children.push(new_child);
+                
+                curr = curr.children.last_mut().unwrap();
+                curr.count += 1;
+                curr_idx = input_ids_len;
+            }
+            
+        }
+    }
+
+    pub fn prefix_match<'a>(&self, input_ids: &'a [usize]) -> &'a [usize] {
+        let mut curr = &self.root;
+
+        let mut curr_idx = 0;
+        let input_ids_len = input_ids.len();
+
+        while curr_idx < input_ids_len {
+            
+            let mut has_full_match = false;
+
+            for i in 0..curr.children.len() {
+                let prefix_len = common_prefix_len(&input_ids[curr_idx..], &curr.children[i].ids);
+                
+                if prefix_len == 0 {
+                    continue;
+                }
+
+                let child = &curr.children[i];
+                
+                if prefix_len == child.ids.len() {
+                    // full match
+                    curr_idx += prefix_len;
+                    curr = child;
+                    has_full_match = true;
+
+                    break;
+                } else {
+                    // partial match
+                    curr_idx += prefix_len;
+                    break;
+                }
+            }
+
+            if !has_full_match {
+                // if not full match, break
+                break;
+            }
+        }
+
+        return &input_ids[..curr_idx];
+
+    }
+
+    pub fn delete(&mut self, input_ids: &[usize]) {
```

Comment:
In the future PRs, we might want to support delete with node contractions.

Reply:
makes sense